File: data structures and algorithms/python/06_070.py
def bucket_sort(arr):
   # Return empty array if input is empty
   if not arr:
      return arr
   
   # Create buckets using array length as bucket count
   max_val, min_val = max(arr), min(arr)
   bucket_count = len(arr)
   buckets = [[] for _ in range(bucket_count)]
   
   # Distribute elements into buckets
   for num in arr:
      # Calculate bucket index for this element
      index = int((num - min_val) * (bucket_count - 1) / 
                 (max_val - min_val))
      buckets[index].append(num)
   
   # Sort individual buckets
   for bucket in buckets:
      bucket.sort()
   
   # Concatenate all buckets into final result
   sorted_arr = []
   for bucket in buckets:
      sorted_arr.extend(bucket)
   
   return sorted_arr

# Example usage
arr = [0.897, 0.565, 0.656, 0.1234, 0.665, 0.3434]
sorted_arr = bucket_sort(arr)
print(sorted_arr)

File: data structures and algorithms/cs/06_070.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class BucketSort
{
   public static void Sort(float[] arr)
   {
      if (arr == null || arr.Length <= 1) return;

      int n = arr.Length;
      List<float>[] buckets = new List<float>[n];

      for (int i = 0; i < n; i++)
      {
         buckets[i] = new List<float>();
      }

      float maxVal = arr.Max();
      float minVal = arr.Min();

      foreach (float num in arr)
      {
         int index = (int)((num - minVal) * (n - 1) / (maxVal - minVal));
         buckets[index].Add(num);
      }

      foreach (var bucket in buckets)
      {
         bucket.Sort();
      }

      int currentIndex = 0;
      foreach (var bucket in buckets)
      {
         foreach (var num in bucket)
         {
            arr[currentIndex++] = num;
         }
      }
   }
}
